pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go

```diff
@@ -0,0 +1,387 @@
+	case err != nil:
+		return err
+	}
+
+	k, err := f.kubeActionsFactory(log, f.env, doc.OpenShiftCluster)
+	if err != nil {
+		return err
+	}
+	e := &etcdrenew{
+		log:           log,
+		k:             k,
+		secretNames:   nil,
+		mode:          "renew",
+		backupSecrets: make(map[string][]byte),
+		lastRevision:  0,
+	}
+
+	if err = e.validateClusterVersion(ctx); err != nil {
+		return err
+	}
+
+	// Fetch secretNames using nodeNames
+	masterNodeNames, err := fetchNodeNames(ctx, k, log)
+	if err != nil {
+		return err
+	}
+	if len(masterNodeNames) != 3 {
+		return api.NewCloudError(http.StatusForbidden, api.CloudErrorCodeForbidden, "", "The cluster doesn't have 3 master nodes")
+	}
+
+	for _, nodeName := range masterNodeNames {
+		for _, prefix := range []string{"etcd-peer-", "etcd-serving-", "etcd-serving-metrics-"} {
+			e.secretNames = append(e.secretNames, prefix+nodeName)
+		}
+	}
+
+	// validate, backup and delete secrets
+	if err = e.validateAndBackup(ctx, true); err != nil {
+		return err
+	}
+
+	if err = e.isRenewed(ctx); err != nil {
+		e.mode = "recovery"
+	} else {
+		e.mode = "renewed"
+	}
+
+	if e.mode == "renewed" {
+		// only validate
+		if err = e.validateAndBackup(ctx, false); err != nil {
+			return err
+		}
+		e.log.Infoln("Done")
+	}
+
+	if e.mode == "recovery" {
+		e.log.Println("Attempting to recover from backup")
+		if err = e.recoverEtcdSecrets(ctx); err != nil {
+			return err
+		}
+		e.log.Infoln("Recovered")
+	}
+
+	return nil
+}
+
+func (e *etcdrenew) validateClusterVersion(ctx context.Context) error {
+	e.log.Infoln("validating cluster version now")
+	rawCV, err := e.k.KubeGet(ctx, "ClusterVersion.config.openshift.io", "", "version")
+	if err != nil {
+		return err
+	}
+	cv := &configv1.ClusterVersion{}
+	err = codec.NewDecoderBytes(rawCV, &codec.JsonHandle{}).Decode(cv)
+	if err != nil {
+		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", fmt.Sprintf("failed to decode clusterversion, %s", err.Error()))
+	}
+	clusterVersion, err := version.GetClusterVersion(cv)
+	if err != nil {
+		return err
+	}
+	// ETCD ceritificates are autorotated by the operator when close to expiry for cluster running 4.9+
+	if !clusterVersion.Lt(version.NewVersion(4, 9)) {
+		return api.NewCloudError(http.StatusForbidden, api.CloudErrorCodeForbidden, "", "etcd certificate renewal is not needed for cluster running version 4.9+")
+	}
+	return nil
+}
+
+func fetchNodeNames(ctx context.Context, k adminactions.KubeActions, log *logrus.Entry) ([]string, error) {
+	var masterNodeNames []string
+	var u unstructured.Unstructured
+	var nodes corev1.NodeList
+
+	log.Infoln("fetching node names")
+
+	nodeList, err := k.KubeList(ctx, "node", "")
+	if err != nil {
+		return nil, err
+	}
+	if err = json.Unmarshal(nodeList, &u); err != nil {
+		return nil, err
+	}
+	err = kruntime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &nodes)
```

Comment:
Why are we converting the response to an unstructured type and then again to a structured type? Can't we just convert directly?